### ISSUE TITLE
Lua Map Support

### ISSUE THREAD

**Issue by SMUnlimited** [SMUnlimited/AMAI#207]
Something to consider feasibility of one day so you don't have to switch to Jass first ... would have to buy reforged first :P

**Comment by jzy-chitong56** [comment]
In China, map authors mostly rely on third-party editors, which inherit many convenient interfaces（Including real-time networking interaction interface and paid unlocking interface）, some of which are implemented through built-in custom libraries. This also means that game ontologies that do not include runtime libraries cannot run these maps, which leads to the inability of Chinese maps to support recast versions. However, Lua can be built into these custom libraries

Additionally, I understand that AI itself does not require Lua, only BJ needs to be converted, but I suspect that the converter may not be able to effectively transpose the BJ of AMAI，Referring to the ready-made Lua file on the w3c map, the declared AI in it is also Jass, so it should be run after compiling and adjusting the transcoder, which may be feasible

**Comment by jzy-chitong56** [comment]
https://github.com/uniqss/jasslua

Not supporting new interfaces, but you can borrow them

**Comment by jzy-chitong56** [comment]
There is one more question

Do you still remember my ITEM ICON? I also put a lot of effort into importing it, but eventually found out that AMAI did not import the list file war3map.imp and war3map.w3t during installation

imp file allows the editor to know which files are imported externally, so that they will still be retained when saving the map

Without this file, AMAI will not destroy the original imported file of the map during import, allowing its original model , sound and effects to function properly. However, AMAI's script will also be lost when saving the map as a result


This causes the AMAI map to lose being edited by the editor ,if editor can change ai script

**Comment by FreeTymeKiyan** [comment]
Wonder if LLM's like ChatGPT can help convert jass to lua more easily

**Comment by SMUnlimited** [comment]
So on LUA maps the commander part runs, and AMAI hits step1 of logging then something goes wrong post that. 
![image](https://github.com/user-attachments/assets/16e9889f-13d3-4c8f-8ff0-61403f196a9d)

**Comment by SMUnlimited** [comment]
Hunted down to the InitCT() method is failing which stores all different characters into a hashtable.

**Comment by SMUnlimited** [comment]
Disabling the language placeholder support seems to work fine from an AI standpoint. Some messages will be lost but a minor loss to address compatibility with both lua and jass at some time. May be able to add support another way or at least add conditional support via another method.

Commander initializes and seems to allow language selection, even prints the intro messages but the keyboard and chat commands do not seem to work.

**Comment by jzy-chitong56** [comment]
> Disabling the language placeholder support seems to work fine from an AI standpoint. Some messages will be lost but a minor loss to address compatibility with both lua and jass at some time. May be able to add support another way or at least add conditional support via another method.
> 
> 
> 
> Commander initializes and seems to allow language selection, even prints the intro messages but the keyboard and chat commands do not seem to work.

https://github.com/jzy-chitong56/map-updater-scripts/tree/master/src/player_features

The W3C keyboard command is written like this. Please take some time to take a look and compare it. He has registered keyboard events for every player once

**Comment by SMUnlimited** [comment]
> > Disabling the language placeholder support seems to work fine from an AI standpoint. Some messages will be lost but a minor loss to address compatibility with both lua and jass at some time. May be able to add support another way or at least add conditional support via another method.
> > Commander initializes and seems to allow language selection, even prints the intro messages but the keyboard and chat commands do not seem to work.
> 
> https://github.com/jzy-chitong56/map-updater-scripts/tree/master/src/player_features
> 
> The W3C keyboard command is written like this. Please take some time to take a look and compare it. He has registered keyboard events for every player once
> 
> In the observer function, he also has string truncation

This appears to be typescsript so not sure that helps unless we have the generated LUA code?
I've confirmed commander is part working as pure Jass within a lua map. So i think there is something else going on. Or its just the triggers that don't work from Jass code.

**Comment by jzy-chitong56** [comment]
[Reading his installation instructions](https://github.com/jzy-chitong56/map-updater-scripts/tree/master/scripts), it seems to explain how TS transcodes to LUA
Alternatively, you can download a W3C client map and directly extract the installed scripts

**Comment by jzy-chitong56** [comment]
https://typescripttolua.github.io/

**Comment by jzy-chitong56** [comment]
look here , W3C map use TS String Split , very very more function

[war3map.zip](https://github.com/user-attachments/files/17431483/war3map.zip)


```
function __TS__StringTrim(self)
    local result = string.gsub(self, "^[%s ﻿]*(.-)[%s ﻿]*$", "%1")
    return result
end

function __TS__StringTrimEnd(self)
    local result = string.gsub(self, "[%s ﻿]*$", "")
    return result
end

function __TS__StringTrimStart(self)
    local result = string.gsub(self, "^[%s ﻿]*", "")
    return result
end

-------------------------------

function __TS__StringSplit(source, separator, limit)
    if limit == nil then
        limit = 4294967295
    end
    if limit == 0 then
        return {}
    end
    local out = {}
    local index = 0
    local count = 0
    if (separator == nil) or (separator == "") then
        while (index < (#source - 1)) and (count < limit) do
            out[count + 1] = __TS__StringAccess(source, index)
            count = count + 1
            index = index + 1
        end
    else
        local separatorLength = #separator
        local nextIndex = (string.find(source, separator, nil, true) or 0) - 1
        while (nextIndex >= 0) and (count < limit) do
            out[count + 1] = __TS__StringSubstring(source, index, nextIndex)
            count = count + 1
            index = nextIndex + separatorLength
            nextIndex = (string.find(
                source,
                separator,
                math.max(index + 1, 1),
                true
            ) or 0) - 1
        end
    end
    if count < limit then
        out[count + 1] = __TS__StringSubstring(source, index)
    end
    return out
end
-----------------------------------------------------
and more
```

**Comment by SMUnlimited** [comment]
Yes in the map scripts you can use lua directly, but not for blizzard.j and ai code which all use jass even in lua mode. So we would have to be able to write lua within the jass files and the games automatic converter doesn't try to convert it somehow.
